Use Brand with  Script Link component on services with variants

Bopchy opened the pull request:
Resolves #4009

**Overall change:** 
_Use Brand with ScriptLink component on services with variants._

**Code changes:**
- _Update Brand version number._
- _Add ScriptLink and SkipLink components to Brand._
- _Update existing snapshots and tests._
- _Update config to include service variants, script link text and script link offscreen text for UKChina, Serbian and Zhongwen._
- _Add unit tests._
- _Add onClick handler that sets the variant preference cookie when the user interacts with the link._
- _Add setPreferredVariantCookie() to UserContext._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [ ] This PR requires manual testing

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into integrate-new-navigation

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into integrate-new-navigation

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into integrate-new-navigation

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into integrate-new-navigation

Commit:
Add variants to config.

Commit:
Update psammead-brand version.

Commit:
Add ScriptLink and Skiplink to Brand. Add variantOffscreenText to config.

Commit:
Update snapshots.

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Add test.

AlistairGempf commented on src/app/containers/Brand/index.jsx:
```diff
@@ -1,22 +1,59 @@
+    script,
+    scriptLinkText,
+    scriptLinkOffscreenText,
+    variant = null,
```
Will this `variant` be the one that the user is already on or the one they would switch to?

Bopchy replied:
It is the one they would switch to

Bopchy replied:
UPDATE: This variant is now the one that the user is currently on

DenisHdz commented on src/app/containers/Brand/index.jsx:
```diff
@@ -1,22 +1,59 @@
+  scriptLinkText,
+  offscreenText,
+) => (
+  <ScriptLink script={script} service={service} href={url} variant={variant}>
```
The service variant preference cookie will be set when a user clicks on the ScriptLink, so we have to create and pass the `onClick` handler to it as a prop. https://github.com/bbc/simorgh/issues/4350#issuecomment-562617433

Bopchy replied:
Added the onClick handler

OlgaLyubin commented on src/app/containers/Brand/index.jsx:
```diff
@@ -29,6 +66,8 @@ const BrandContainer = props => {
       maxWidth={maxWidth}
       svg={brandSVG}
       url={`/${service}`}
+      scriptLink={scriptLink && scriptLink}
```
What is the reason for having two `scriptLink` here?

Bopchy replied:
I had previously added it as a check, so that `scriptLink` is never empty. But this is not required. Removed the second `scriptLink`

Commit:
Rename variant to scriptLinkVariant. Make scriptLinkVariant the variant to be switched to.

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Add setCookie() to ScriptLink.

Commit:
Add comments.

Commit:
Fix setCookie() not working correctly.

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Add tests.

tochwill commented on src/app/containers/Brand/index.jsx:
```diff
@@ -1,22 +1,70 @@
+import { setPreferredVariantCookie } from '#contexts/UserContext/cookies';
 import { ServiceContext } from '#contexts/ServiceContext';
 
+const setCookie = (service, scriptLinkVariant) => {
```
Do we need this `setCookie` function, or could you just used `setPreferredVariantCookie` directly?

Bopchy replied:
`setPreferredVariantCookie` can be used directly. Switched to using it directly

tochwill commented on src/app/containers/Brand/index.jsx:
```diff
@@ -1,22 +1,70 @@
+
+const renderSkipLink = (service, script) => (
+  <SkipLink service={service} script={script} href="#content">
+    Skip to content
```
Does this need translating?

Bopchy replied:
Yes, it does. This now gets translated.

tochwill commented on src/app/containers/Brand/index.jsx:
```diff
@@ -1,22 +1,70 @@
+    variant={scriptLinkVariant}
+    onClick={() => setCookie(service, scriptLinkVariant)}
+  >
+    <span aria-hidden>{scriptLinkText}</span>
```
I think this needs to be `aria-hidden="true"`, as the aria tag expects a string not a boolean value!

Bopchy replied:
Made this `aria-hidden="true"`

tochwill commented on src/app/containers/Brand/index.test.jsx:
```diff
@@ -42,5 +61,20 @@ describe(`BrandContainer`, () => {
       expect(skipLinkHref).toBe('#content');
     });
+
+    it('should update cookie to preffered variant when ScriptLink is clicked', () => {
```
Do we also need a test for when a preference cookie has already been set?

Bopchy replied:
Added a test for when a preference cookie already exists.

tochwill commented on src/app/contexts/UserContext/cookies/index.js:
```diff
@@ -6,12 +6,21 @@ export const getCookiePolicy = () => {
   return Cookie.get(POLICY_COOKIE) || '000';
 };
 
+export const getPreferredVariantCookieName = service => `ckps_${service}`;
```
Is this exported function used anywhere, or could you just set the cookie to `ckps_${service}` below (in the same way as `VARIANT_COOKIE` is set in `getPreferredVariant`?

Bopchy replied:
`getPreferredVariantCookieName` was not being used elsewhere. Updated `setPreferredVariantCookie` to determine the variant cookie name as suggested

tochwill reviewed (commented):
Looks really good! Just have a few questions :)

Commit:
Remove setCookie(), use setPreferredVariantCookie(). Remove getPreferredVariantCookieName().

Commit:
Add skipLinkText translations. Update snapshots. Update tests.

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Update serbian, ukchina, zhongwen configs. Create getOtherVariant(), useCurrentVariant().

Commit:
Update renderScriptLink().

Commit:
Get setPreferredVariantCookie from UserContext.

Commit:
Update tests. Remove useParams() hook.

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Add tests.

sadickisaac commented on src/app/containers/Brand/index.test.jsx:
```diff
@@ -42,5 +77,42 @@ describe(`BrandContainer`, () => {
+
+      beforeEach(() => {
+        container = render(
+          BrandContainerWithContext({
```
```suggestion
          BrandContainerWithContext(variantServiceContextStub)
```

sadickisaac commented on src/app/containers/Brand/index.test.jsx:
```diff
@@ -17,11 +20,38 @@ const newsServiceContextStub = {
+
+variantHandlers.getOtherVariant.mockImplementation(() => 'test');
+
+const spy = jest.spyOn(cookies, 'setPreferredVariantCookie');
```
While naming this `spy` would be appropriate if it was in close proximity to where it was being used. Scrolling down to where `spy` is used requires one to keep the information that it points to `setPreferredVariantCookie` in his/her head. Could we rename this to `setPreferredVariantCookieSpy`.

Bopchy replied:
Yes. Renamed this to `setPreferredVariantCookieSpy` as suggested.

sadickisaac commented on src/app/containers/Brand/index.test.jsx:
```diff
@@ -42,5 +77,42 @@ describe(`BrandContainer`, () => {
+
+        expect(spy).toHaveBeenCalledTimes(1);
+        expect(document.cookie).toEqual('; ckps_news=test');
+      });
```
The two tests are common since they are checking for the update of `cookie`. Could we combine this to one test instead of two.

Bopchy replied:
I think it makes sense to leave the test as it is right now because it is more readable. Currently as it is, its very clear that clicking on the ScriptLink should set cookie when one is not defined and when one is defined - which is slightly harder to see if it's all one test

sadickisaac commented on src/app/containers/Brand/index.test.jsx:
```diff
@@ -42,5 +77,42 @@ describe(`BrandContainer`, () => {
+          BrandContainerWithContext({
+            ...variantServiceContextStub,
+          }),
+        ).container;
```
Could we also set `scriptLink` here
```suggestion
scriptLink = container.querySelector('a[data-variant="test"]');
```

sadickisaac commented on src/app/lib/config/services/serbian.js:
```diff
@@ -184,6 +184,16 @@ const baseServiceConfig = {
+      scriptLinkText: 'Ћир',
+      scriptLinkOffscreenText: 'Cyrillic',
+    },
+  },
```
Since we already have the `lat` and `cyr` keys on the service config, it makes sense to place the `scripLink` stuff inside their respective `keys` instead of repeating them here.
Something like 
```js
export const service = {
  lat: {
    ...baseServiceConfig,
    skipLink: {
      text: 'Lat',
      offscreenText: 'Latin'
    }
  },
  cyr: {
    ...baseServiceConfig,
    skipLink: {
      text: 'Ћир',
      offscreenText: 'Cyrillic'
    }
  },
}

Bopchy replied:
The challenge I had with structuring this as you've suggested is that I was not able to easily access the scriptLink data of the opposite variant - as the data passed through ServiceContext is variant specific. So, if the currentVariant is `lat`, I couldn't access ```skipLink: {
      text: 'Ћир',
      offscreenText: 'Cyrillic'
    }```

Bopchy replied:
UPDATE: Paired with @sadickisaac on this, and adopted the suggestion.

Commit:
Apply suggestions from code review.

Co-Authored-By: sadickisaac <48688870+sadickisaac@users.noreply.github.com>

Commit:
Apply suggestions from code review

Co-Authored-By: sadickisaac <48688870+sadickisaac@users.noreply.github.com>

Commit:
Merge branch '4009-use-brand-with-script-link' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Update serbian, ukchina, zhongwen configs. Update scriptLink switching functionality.
- Update ServiceContextProvider to include prefferedServiceVariant and otherVariant.
- Update withContext to update scriptLink to opposite variant.

Commit:
Update test.

OlgaLyubin reviewed (approved):
Looks good except that Jenkins is failing over bundle sizes 👍

sadickisaac commented on src/app/lib/config/services/serbian.js:
```diff
@@ -234,6 +239,11 @@ export const service = {
     script: cyrillicAndLatin,
     datetimeLocale: `sr-cyrl`,
+    scriptLink: {
+      variant: 'cyr',
```
`scriptLink` is already under the `cyr` variant key in the service config. Can we remove this here.

Bopchy replied:
Removed the `variant` attribute.

Commit:
Use RequestContext.

Commit:
Undo .env changes.

Commit:
Merge branch '4009-use-brand-with-script-link' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Update test.

Bopchy commented:
Updated to latest after being prompted to do so locally. This ended up pulling in all of the additional commits from latest

Commit:
Revert "Use RequestContext."

This reverts commit 7630067afb8874ee5068737b2dc717275437f288.

Commit:
Update renderScriptLink

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Update variant configs.

Commit:
Update bundleSizeConfig.

Commit:
Update e2es.

Bopchy commented:
Travis bbc-a11y failures are being caused by `scrollable-nav` and `menu button` changes introduced in the [integrate-new-navigation](https://github.com/bbc/simorgh/tree/integrate-new-navigation) branch

PriyaKR commented:
Testing:

Noticed few issues across cross browser and device testing

- On desktop browsers when on http://localhost:7080/ukchina/trad and then click one of the links on the top nav and then click the brand svg it gets redirected to http://localhost:7080/ukchina/trad but for a short time http://localhost:7080/ukchina/simp url is seen
- On mobile browsers when on http://localhost:7080/ukchina/trad and then click one of the links on the top nav and then click the brand svg it gets redirected to http://localhost:7080/ukchina/simp and not to http://localhost:7080/ukchina/trad
- Also on IE9 the service variant button appears to be in wrong place just below the brand svg not an issue on test env.

Bopchy commented:
```On desktop browsers when on http://localhost:7080/ukchina/trad and then click one of the links on the top nav and then click the brand svg it gets redirected to http://localhost:7080/ukchina/trad but for a short time http://localhost:7080/ukchina/simp url is seen``` should be resolved by https://github.com/bbc/simorgh/issues/4976. 

Seeing `http://localhost:7080/ukchina/simp` for a short time before it redirects to `http://localhost:7080/ukchina/trad` is being caused by the default variant being accessed before the cookie redirect happens

Bopchy commented:
Regarding ```Also on IE9 the service variant button appears to be in wrong place just below the brand svg not an issue on test env.```, the service variant button appears just below the brand svg only on small screen widths on IE, which is in-line with the designs

<img width="452" alt="Screenshot 2020-01-01 at 23 09 17" src="https://user-images.githubusercontent.com/12303856/71645617-e6d6e080-2ceb-11ea-8351-ade8afe0bcf2.png">

PriyaKR commented:
@Bopchy i see the IE9 issue happening on wider screens.

Bopchy commented:
The IE9 issue (pictured below) is happening because IE9 doesn't fully support `display: flex` which has been used to style the `div.SvgWrapper`, which contains all the `<Banner />` links. As such the contents of `display: flex` are not properly positioned out of the box. This has been documented [here](https://github.com/philipwalton/flexbugs/issues/261). 

<img width="637" alt="Screenshot 2020-01-03 at 00 00 25" src="https://user-images.githubusercontent.com/12303856/71692899-397fcd80-2dbc-11ea-8011-19accf22279a.png">

Will not be fixing this though as IE9 is not on the list of [supported browsers](https://github.com/bbc/psammead#browser-support)

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Update snapshots.

Commit:
Revert scriptLink config editing.

Commit:
Remove reassignment.

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Bopchy commented:
```On mobile browsers when on http://localhost:7080/ukchina/trad and then click one of the links on the top nav and then click the brand svg it gets redirected to http://localhost:7080/ukchina/simp and not to http://localhost:7080/ukchina/trad``` is happening because the preferred variant cookie is not being set on mobile - so clicking on the brand svg redirects to the default variant

This should be fixed in https://github.com/bbc/simorgh/issues/5038

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Commit:
Merge branch 'integrate-new-navigation' of github.com:bbc/simorgh into 4009-use-brand-with-script-link

Bopchy commented:
Travis failures will be fixed in https://github.com/bbc/simorgh/tree/integrate-new-navigation once this work is merged, as the same failures are happening in parent branch as well

PriyaKR commented:
@Bopchy I am now seeing the mobile issue mentioned above on desktop as well.May be we should tweak the issue saying `Cookie redirects do not work on mobile and desktop`

PriyaKR commented:
I have updated the issue https://github.com/bbc/simorgh/issues/5038 to include the desktop issue as well.So this PR can be merged now as separate issue is being created for the bugs mentioned above.

tochwill committed:
Merge branch 'integrate-new-navigation' into 4009-use-brand-with-script-link

tochwill committed:
Update snapshots
